packages/pg/lib/sasl.js

```diff
@@ -20,19 +20,27 @@ function continueSession(session, password, serverData) {
   if (session.message !== 'SASLInitialResponse') {
     throw new Error('SASL: Last message was not SASLInitialResponse')
   }
+  if (typeof password !== 'string') {
+    throw new Error('SASL: SCRAM-SERVER-FIRST-MESSAGE: client password must be a string')
+  }
+  if (typeof serverData !== 'string') {
+    throw new Error('SASL: SCRAM-SERVER-FIRST-MESSAGE: serverData must be a string')
+  }
 
-  const sv = extractVariablesFromFirstServerMessage(serverData)
+  const sv = parseServerFirstMessage(serverData)
 
   if (!sv.nonce.startsWith(session.clientNonce)) {
     throw new Error('SASL: SCRAM-SERVER-FIRST-MESSAGE: server nonce does not start with client nonce')
+  } else if (sv.nonce.length === session.clientNonce.length) {
+    throw new Error('SASL: SCRAM-SERVER-FIRST-MESSAGE: server nonce is too short')
```

Comment:
nit: could these two if conditions just be combined into `if (sv.nonce !== session.clientNonce)`? As I think if a string starts with another string and has exactly the same length then they're the same string?

Reply:
No it's two separate checks. The RFC, https://tools.ietf.org/html/rfc5802, says:

> In response, the server sends a "server-first-message" containing the
> user's iteration count i and the user's salt, and appends its own
> nonce to the client-specified one.

So the server's response must be the clients nonce plus some random server nonce. Need to ensure our original client nonce is there (which the previous code was already doing). The additional check is to ensure the server added something non-empty of it's own. Separate checks gives a bit prettier error message if something is off.

Reply:
ohhh that makes sense yes you're right I actually read that backwards when I was reviewing - if the length is equal it __is__ a problem, not if the length is equal it is not a problem. Got it.